pkg/cassandra/util.go

```diff
@@ -31,14 +33,32 @@ func DatacenterStopping(dc *cassdcapi.CassandraDatacenter) bool {
 	return dc.GetConditionStatus(cassdcapi.DatacenterStopped) == corev1.ConditionTrue && dc.Status.CassandraOperatorProgress == cassdcapi.ProgressUpdating
 }
 
-func ComputeSystemReplication(kluster *api.K8ssandraCluster) SystemReplication {
+func GetDatacentersForReplication(kc *api.K8ssandraCluster) []api.CassandraDatacenterTemplate {
+	var datacenters []api.CassandraDatacenterTemplate
+	if kc.Status.GetConditionStatus(api.CassandraInitialized) == corev1.ConditionTrue {
+		datacenters = make([]api.CassandraDatacenterTemplate, 0)
+		for _, dc := range kc.Spec.Cassandra.Datacenters {
+			if status, found := kc.Status.Datacenters[dc.Meta.Name]; found && status.Cassandra.GetConditionStatus(cassdcapi.DatacenterReady) == corev1.ConditionTrue {
+				datacenters = append(datacenters, dc)
+			}
+		}
+	} else {
+		datacenters = kc.Spec.Cassandra.Datacenters
```

Comment:
I'm slightly confused here, why do we want only Ready ones if initialized is true, but everything if Initialized is false? Feels unnatural why this path should ever be taken instead of just returning nothing.

Reply:
Or is this path taken when we initialize all the datacenters the first time K8ssandraCluster is created?

Reply:
>Or is this path taken when we initialize all the datacenters the first time K8ssandraCluster is created?

short answer - yes

long answer...

We configure replication of system keyspaces by setting a couple system properties:

* `cassandra.system_distributed_replication_dc_names`
* `cassandra.system_distributed_replication_per_dc`

management-api looks for these properties and applies changes to the system_auth keyspace using internal Cassandra APIs, not CQL. If we are creating a cluster with 2 dcs, dc1 and dc2, both will be configured for system_auth.   Note that what I am describing is already the current behavior in main.

In 4.0 Cassandra won't let you specify the replication for a keyspace that does not exist (at least through CQL). I had to introduce changes in this PR for configuring the replication of the Stargate auth and Reaper keyspaces. Those are configured via CQL calls. So going back to the 2 DCs example, when dc1 becomes ready we can create the keyspaces and specify replication for dc1 but not for dc2. Cassandra won't allow this.  This resulted in different handling of system keyspaces vs other keyspaces.

Once I added support for handling non-system keyspaces, I attempted (in vain) to handle system keyspaces the same way to make the code less confusing and more maintainable. There are queries with a `LOCAL_ONE` consistency executed in dc2 that will fail with auth error because the replica(s) in dc2 need a repair.

In summary, until we implement support for running repairs, at least of system_auth, we have to handle system keyspaces the way we currently do. That means 1) for system keyspaces we need to specify replication for all DCs up front and 2) for non-system keyspaces we cannot specify the replication for all DCs up front.

I will rename the method and add some docs to make it more clear that it is only intended for system keyspaces.